### DISCUSSION TITLE
Using Pocketbase in NODE

### DISCUSSION THREAD

**Discussion by Arinji2** [pocketbase/pocketbase#3688] [Q&A] [answered]
Hiya, so I was trying to setup pocketbase in a node express app. 
I followed this to get node express and nodemon with ts setup [https://mpolinowski.github.io/docs/Development/Javascript/2023-06-04-nodejs-typescript-sql/2023-06-04/#nodejs](url)

It works but whenever i try to use pocketbase i get this error
`
Error [ERR_REQUIRE_ESM]: require() of ES Module C:\Users\arinj\OneDrive\Desktop\Coding\Arinji\vibeify-api\node_modules\pocketbase\dist\pocketbase.es.mjs not supported.
Instead change the require of C:\Users\arinj\OneDrive\Desktop\Coding\Arinji\vibeify-api\node_modules\pocketbase\dist\pocketbase.es.mjs to a dynamic 
import() which is available in all CommonJS modules.
    at Object.<anonymous> (C:\Users\arinj\OneDrive\Desktop\Coding\Arinji\vibeify-api\src\index.ts:7:38)
    at m._compile (C:\Users\arinj\OneDrive\Desktop\Coding\Arinji\vibeify-api\node_modules\ts-node\dist\index.js:857:29)
    at require.extensions.<computed> [as .ts] (C:\Users\arinj\OneDrive\Desktop\Coding\Arinji\vibeify-api\node_modules\ts-node\dist\index.js:859:16) 
    at phase4 (C:\Users\arinj\OneDrive\Desktop\Coding\Arinji\vibeify-api\node_modules\ts-node\dist\bin.js:466:20)
    at bootstrap (C:\Users\arinj\OneDrive\Desktop\Coding\Arinji\vibeify-api\node_modules\ts-node\dist\bin.js:54:12)
    at main (C:\Users\arinj\OneDrive\Desktop\Coding\Arinji\vibeify-api\node_modules\ts-node\dist\bin.js:33:12)
    at Object.<anonymous> (C:\Users\arinj\OneDrive\Desktop\Coding\Arinji\vibeify-api\node_modules\ts-node\dist\bin.js:579:5) {
  code: 'ERR_REQUIRE_ESM'
}`

Things i have tried
1) Using the require pocketbase/cjs. But it dosent have types which is hard to work without
2) Using "type": "module" in package.json.. Adding that breaks the appp `
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for C:\Users\arinj\OneDrive\Desktop\Coding\Arinji\vibeify-api\src\index.ts`

src/index.ts
```js
import express from "express";
import Pocketbase from "pocketbase";

const app = express();

const PORT = 3000;

app.get("/", (req, res) => {
  res.send("Hello world!");
  const pb = new Pocketbase("test");
});

app.listen(PORT, () => {
  console.log(`Server listening on port ${PORT}`);
});
```

package.json
```json
{
  "name": "vibeify-api",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "start": "ts-node src/index.ts",
    "dev": "nodemon src/index.ts"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@types/express": "^4.17.20",
    "nodemon": "^3.0.1",
    "ts-node": "^10.9.1",
    "typescript": "^5.2.2"
  },
  "dependencies": {
    "express": "^4.18.2",
    "pocketbase": "^0.19.0"
  }
}
```

**Comment by ganigeorgiev** [comment]
The JS SDK is a ES module.

The error is most likely caused by the TS config and you'll get the same for any ESM-first npm package.

If you've directly followed the tutorial from the above link, you can try updating your tsconfig and change the `module` to  `nodenext` or remove it to use the default. From the [TypeScript documentation](https://www.typescriptlang.org/tsconfig#module):

> Module - CommonJS if [target](https://www.typescriptlang.org/tsconfig#target) is ES3 or ES5; **ES6/ES2015 otherwise**.

_(in the above tutorial `target` seems to be set to `es2016`; you can also change the `target` prop to `esnext`)_

**Reply by Arinji2** [reply]
Alright so changed the ts config thingy to nodenext.. having the modules thing still shows the same error.. removing it gives this

`The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is 
an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("pocketbase")' call instead.`

**Reply by Arinji2** [reply]
tried changing it to esnext and kept the modules thing, as long as i have the "type": "modules" in the package.json i seem to still be getting the unknow file extension .ts

**Reply by ganigeorgiev** [reply, marked as answer]
Try changing your package.json `tsnode` script and pass the `--esm` flag (or use `ts-node-esm`; see https://www.npmjs.com/package/ts-node#esm), aka. `"start": "ts-node --esm src/index.ts"`.

If that doesn't work, then I'm not sure what could be the cause of the error but I don't think it is PocketBase related. You can try contacting the author of the tutorial and see if they can recommend something else.
